### PR TITLE
Add checks to convert std library iterator algorithms into c++20 or boost ranges

### PR DESCRIPTION
**Pull request by njames93**

### CODE REVIEW

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+namespace clang::tidy::utils {
+
+static bool operator==(const UseRangesCheck::Replacer::Indexes &L,
+                       const UseRangesCheck::Replacer::Indexes &R) {
+  return std::tie(L.BeginArg, L.EndArg, L.ReplaceArg) ==
+         std::tie(R.BeginArg, R.EndArg, R.ReplaceArg);
+}
+
+std::string
+getFullPrefix(ArrayRef<UseRangesCheck::Replacer::Indexes> Signature) {
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:47]: Nit: static, keep this private

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  return llvm::hash_combine(L.BeginArg, L.EndArg, L.ReplaceArg);
+}
+
+namespace {
+
+AST_MATCHER(Expr, hasSideEffects) {
+  return Node.HasSideEffects(Finder->getASTContext());
+}
+
+AST_MATCHER_P(Expr, isEquivalentToBound, std::string, Other) {
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:67]: there is already something similar: clang::tidy::matchers::isStatementIdenticalToBoundNode, but that one by default pass false as Canonical

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+                ast_matchers::internal::DynTypedMatcher::VO_AnyOf,
+                ASTNodeKind::getFromNodeKind<CallExpr>(),
+                std::move(TotalMatchers))
+                .convertTo<CallExpr>()),
+        this);
+  }
+}
+
+// Returns the proper token based end location of \p E.
+static SourceLocation exprLocEnd(const Expr *E, const ASTContext &Context) {
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:152]: Consider moving it to LexerUtils

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+      for (const auto &ArgPair : Signature) {
+        Matchers.push_back(makeMatcherPair(getFullPrefix(Signature), ArgPair));
+      }
+      TotalMatchers.push_back(
+          ast_matchers::internal::DynTypedMatcher::constructVariadic(
+              ast_matchers::internal::DynTypedMatcher::VO_AllOf,
+              ASTNodeKind::getFromNodeKind<CallExpr>(), std::move(Matchers)));
+    }
+    Finder->addMatcher(
+        callExpr(
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:140]: consider using TK_IgnoreUnlessSpelledInSource, to exclude some implicit code, like template instances.
Just to reduce risk of fix-conflicts and improve check performance.

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "../ClangTidyCheck.h"
+#include "IncludeInserter.h"
+#include "clang/AST/Decl.h"
+#include "llvm/ADT/IntrusiveRefCntPtr.h"
+#include "llvm/ADT/StringMap.h"
+
+namespace clang::tidy::utils {
+
+/// FIXME: Base class for handling converting std iterator algorithms to a range
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.h:20]: Nit: Maybe just drop "FIXME"

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+/// equivalent
+class UseRangesCheck : public ClangTidyCheck {
+public:
+  class Replacer : public llvm::RefCountedBase<Replacer> {
+  public:
+    struct Indexes {
+      enum Replace { First, Second };
+      unsigned BeginArg;
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.h:28]: Nit: `BeginArg = 0U`,

File: clang-tools-extra/docs/clang-tidy/checks/boost/use-ranges.rst [hunk 7]
```diff
@@ -0,0 +1,24 @@
+.. title:: clang-tidy - boost-use-ranges
+
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/boost/use-ranges.rst:2]: add options that check has (like include style).

Add table with info what is converted to what.

File: clang-tools-extra/docs/clang-tidy/checks/boost/use-ranges.rst [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+boost-use-ranges
+================
+
+Detects calls to standard library iterator algorithms that could be replaced
+with a ranges version instead
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/boost/use-ranges.rst:7]: Nit: '.' after instead

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-ranges.rst [hunk 9]
```diff
@@ -0,0 +1,24 @@
+.. title:: clang-tidy - modernize-use-ranges
+
+modernize-use-ranges
+====================
+
+Detects calls to standard library iterator algorithms that could be replaced
+with a ranges version instead
+
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-ranges.rst:8]: Same:
- table with list of what is converted to what
- check options

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-ranges.rst [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  auto AreSame = std::equal(std::execution::par, Items1.cbegin(), Items1.cend(),
+                            std::begin(Items2), std::end(Items2));
+
+
+transforms to:
+
+.. code-block:: c++
+
+  auto Iter1 = std::ranges::find(Items, 0);
+  auto AreSame = std::equal(std::execution::par, Items1, Items2);
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-ranges.rst:24]: should be "std::ranges::equal"

File: clang-tools-extra/test/clang-tidy/checkers/boost/use-ranges.cpp [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+  // CHECK-FIXES: boost::algorithm::is_sorted_until(I);
+
+  std::reduce(I.begin(), I.end());
+  // CHECK-MESSAGES: :[[@LINE-1]]:3: warning: Use a ranges version of this algorithm
+  // CHECK-FIXES: boost::algorithm::reduce(I);
+
+  std::reduce(I.begin(), I.end(), 2);
+  // CHECK-MESSAGES: :[[@LINE-1]]:3: warning: Use a ranges version of this algorithm
+  // CHECK-FIXES: boost::algorithm::reduce(I, 2);
+}
```
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/boost/use-ranges.cpp:132]: add test with:
-  cbegin/cend
- lambda

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-ranges.cpp [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+  // CHECK-FIXES: std::ranges::find(I, 5);
+}
+
+void Negatives() {
+  std::vector<int> I, J;
+  std::find(I.begin(), J.end(), 0);
+  std::find(I.begin(), I.begin(), 0);
+  std::find(I.end(), I.begin(), 0);
+  std::equal(I.begin(), J.end(), I.begin(), I.end());
+}
```
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/modernize/use-ranges.cpp:171]: add some tests with lambdas (custom comparator / predicate)

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+  auto Iter = Replaces.find(QN);
+  assert(Iter != Replaces.end());
+  SmallString<64> Buffer;
+  for (const auto &Signature : Iter->getValue()->getReplacementSignatures()) {
+    Buffer.assign({BoundCall, getFullPrefix(Signature)});
+    const auto *Call = Result.Nodes.getNodeAs<CallExpr>(Buffer);
+    if (!Call)
+      continue;
+    auto Diag =
+        diag(Call->getBeginLoc(), "Use a ranges version of this algorithm");
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:199]: "Use a ranges version of this algorithm" -> "use a ranges version of this algorithm"

**Review by PiotrZSL — COMMENTED** [review]
Few nits. Mainly documentation require some love.

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp (hunk 3 above)
Comment (njames93) [reply clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp]: I just removed this call, had one call site and just inlined instead

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-ranges.rst (hunk 9 above)
Comment (PiotrZSL) [reply clang-tools-extra/docs/clang-tidy/checks/modernize/use-ranges.rst:8]: Table/list anything, just so user would know what is actually being replaced by check.

File: clang-tools-extra/clang-tidy/boost/UseRangesCheck.h [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+class UseRangesCheck : public utils::UseRangesCheck {
+public:
+  UseRangesCheck(StringRef Name, ClangTidyContext *Context);
+
+  void storeOptions(ClangTidyOptions::OptionMap &Options) override;
+
+  ReplacerMap getReplacerMap() const override;
+
+  DiagnosticBuilder createDiag(const CallExpr &Call) override;
+
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/boost/UseRangesCheck.h:29]: Shouldn't `isLanguageVersionSupported` be implemented?

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+                               llvm::SmallVectorImpl<FixItHint> &Output,
+                               const ASTContext &Ctx) {
+  llvm::SmallVector<unsigned> Sorted(Indexes);
+  // Keep track of commas removed
+  llvm::SmallBitVector Commas(Call.getNumArgs());
+  // The first comma is actually the '(' which we can't remove
+  Commas[0] = true;
+  llvm::sort(Sorted);
+  for (auto Index : Sorted) {
+    const auto *Arg = Call.getArg(Index);
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:154]: `auto` should not be used - type is not spelled explicitly.

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+      Output.push_back(FixItHint::CreateRemoval(CharSourceRange::getTokenRange(
+          Arg->getBeginLoc().getLocWithOffset(-1), Arg->getEndLoc())));
+      Commas[Index] = true;
+    }
+  }
+}
+
+void UseRangesCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *Function = Result.Nodes.getNodeAs<FunctionDecl>(FuncDecl);
+  auto QN = "::" + Function->getQualifiedNameAsString();
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:179]: Ditto.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 17]
```diff
@@ -131,6 +131,12 @@ Improvements to clang-tidy
 New checks
 ^^^^^^^^^^
 
+- New :doc:`boost-use-ranges
+  <clang-tidy/checks/boost/use-ranges>` check.
+
+  Detects calls to standard library iterator algorithms that could be replaced
+  with a boost ranges version instead
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:138]: Missing period. Same in other description.

File: clang-tools-extra/clang-tidy/boost/UseRangesCheck.h (hunk 14 above)
Comment (njames93) [reply clang-tools-extra/clang-tidy/boost/UseRangesCheck.h:30]: It's not necessary, the base `utils::UseRangesCheck` already requires C++11

**Review by 5chmidti — COMMENTED** [review]
This will be great to have. 

It should be possible to add support for template function decls in a follow-up. Maybe this should be documented to avoid confusion?

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.h [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+/// Base class for handling converting std iterator algorithms to a range
+/// equivalent.
+class UseRangesCheck : public ClangTidyCheck {
+public:
+  struct Indexes {
+    enum Replace { First, Second };
+    unsigned BeginArg;
+    unsigned EndArg = BeginArg + 1;
+    Replace ReplaceArg = First;
+  };
```
[the remark this answers is not in the document]
Comment (njames93) [reply clang-tools-extra/clang-tidy/utils/UseRangesCheck.h:31]: It was added to support downstream users who want to extend the check to work for their own custom libraries who's methods are a little different to those found in the standard library.

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+                     hasDeclaration(functionDecl(
+                         IsBegin ? hasAnyName("::std::begin", "::std::cbegin")
+                                 : hasAnyName("::std::end", "::std::cend"))))));
+}
+static ast_matchers::internal::Matcher<CallExpr>
+makeMatcherPair(StringRef State, const UseRangesCheck::Indexes &Indexes) {
+  auto ArgPostfix = std::to_string(Indexes.BeginArg);
+  SmallString<64> ID = {BoundCall, State};
+  return callExpr(argumentCountAtLeast(
+                      std::max(Indexes.BeginArg, Indexes.EndArg) + 1),
```
[the remark this answers is not in the document]
Comment (njames93) [reply clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:93]: Again to support any weird libraries downstream users may have.

File: clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+
+static auto makeMatcher(bool IsBegin, StringRef Prefix) {
+  auto Member =
+      IsBegin ? expr(unless(hasSideEffects())).bind((ArgName + Prefix).str())
+              : expr(matchers::isStatementIdenticalToBoundNode(
+                    (ArgName + Prefix).str()));
+  return expr(
+      anyOf(cxxMemberCallExpr(
+                callee(cxxMethodDecl(IsBegin ? hasAnyName("begin", "cbegin")
+                                             : hasAnyName("end", "cend"))),
```
[the remark this answers is not in the document]
Comment (njames93) [reply clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp:81]: How will that play with the c++23 `deducing this` feature?

File: clang-tools-extra/docs/ReleaseNotes.rst (hunk 17 above)
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:138]:
```suggestion
  with a Boost ranges version instead.
```

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 21]
```diff
@@ -169,6 +175,12 @@ New checks
   Finds initializer lists for aggregate types that could be
   written as designated initializers instead.
 
+- New :doc:`modernize-use-ranges
+  <clang-tidy/checks/modernize/use-ranges>` check.
+
+  Detects calls to standard library iterator algorithms that could be replaced
+  with a ranges version instead
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:182]:
```suggestion
  with a ranges version instead.
```

File: clang-tools-extra/docs/clang-tidy/checks/boost/use-ranges.rst [hunk 22]
```diff
@@ -0,0 +1,54 @@
+.. title:: clang-tidy - boost-use-ranges
+
+boost-use-ranges
+================
+
+Detects calls to standard library iterator algorithms that could be replaced
+with a boost ranges version instead.
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/boost/use-ranges.rst:7]:
```suggestion
with a Boost ranges version instead.
```

File: clang-tools-extra/docs/clang-tidy/checks/boost/use-ranges.rst [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+.. option:: IncludeStyle
+
+   A string specifying which include-style is used, `llvm` or `google`. Default
+   is `llvm`.
+
+.. option:: IncludeBoostSystem
+   
+   If `true` the boost headers are included as system headers with angle
+   brackets (`#include <boost.hpp>`), otherwise quotes are used
+   (`#include "boost.hpp"`).
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/boost/use-ranges.rst:54]: Missing default value.

**Comment by njames93** [comment]
- Added some more functions for boost/range/numeric
- Added support for Reverse range iteration

**Review by PiotrZSL — APPROVED** [review]
Looks good enough for me.

Note: Prefer pushing changes in additional commits instead of merge + rebase, it's easier to track what changed between first review and fixes.

**Comment by njames93** [comment]
I no longer have commit access, would you do the honours instead

**Comment by PiotrZSL** [comment]
@njames93 You should be able to merge, if not get, commit access renewed, after all you still an owner of this tool.

**Comment by HighCommander4** [comment]
FYI, the new `boost-use-ranges` check is pretty crashy in clangd (https://github.com/llvm/llvm-project/issues/109037, https://github.com/llvm/llvm-project/issues/109367, https://github.com/clangd/clangd/issues/2173, https://github.com/clangd/clangd/issues/2151).

**Comment by hia3** [comment]
It crashes most likely because a local variable is captured in the static lambda. https://github.com/llvm/llvm-project/blob/bf895c714e1f8a51c1e565a75acf60bf7197be51/clang-tools-extra/clang-tidy/boost/UseRangesCheck.cpp#L208

**Comment by HighCommander4** [comment]
> It crashes most likely because a local variable is captured in the static lambda.
> 
> https://github.com/llvm/llvm-project/blob/bf895c714e1f8a51c1e565a75acf60bf7197be51/clang-tools-extra/clang-tidy/boost/UseRangesCheck.cpp#L208

Nice find! That does seem to be the problem. I've put up a fix at https://github.com/llvm/llvm-project/pull/111282.

**Pull request merged**